### PR TITLE
Block creating long table names in JDBC and MongoDB connectors

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Relates to #12882

## Documentation

(x) No documentation is needed.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# PostgreSQL
* Prevent creating a new table which the table name is longer than the max length. 
  Previously, the name was silently truncated to the max length (63 characters by default). ({issue}`12892`)

# MongoDB
* Create a collection explicitly when creating a new table. 
  Previously, it was created when the data is written for the first time. ({issue}`12892`)
```

### CODE REVIEW

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java [hunk 1]
```diff
@@ -555,6 +557,8 @@ protected JdbcOutputTableHandle createTable(ConnectorSession session, ConnectorT
             String remoteTargetTableName = identifierMapping.toRemoteTableName(identity, connection, remoteSchema, targetTableName);
             String catalog = connection.getCatalog();
 
+            verifyTableNameLength(connection.getMetaData(), remoteTargetTableName);
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:560]: We shouldn't need to do this here, so we shouldn't do this.

We generally trust that CREATE TABLE statement will do all the proper validation on the remote database's side.

If we cannot trust the database fully, why do we trust `databaseMetadata.getMaxTableNameLength`? It can be incorrect as well. 
(a driver has smaller user base than the database engine itself, and metadata methods may get even less scrutiny...)

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 2]
```diff
@@ -1978,6 +1978,38 @@ public void testCreateTable()
...
+    @Test
+    public void testCreateTableWithLongTableName()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_CREATE_TABLE));
+
+        String baseTableName = "test_create_" + randomTableSuffix();
+
+        if (maxTableNameLength().isEmpty()) {
+            // assume 65535 is enough for most use cases
+            String tableName = baseTableName + "z".repeat(65535);
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:1990]: This creates name longer than 65535. Intentional? Comment suggests it isn't.

the `else` branch takes care to create name of precisely given length, so i would unify them with
```
int maxLength = maxTableNameLength()
        // Assume 2^16 is enough for most use cases. Add a bit more to ensure 2^16 isn't actual limit.
        .orElse(65536 + 5);
```

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 3]
```diff
@@ -1978,6 +1978,38 @@ public void testCreateTable()
...
+            assertTrue(getQueryRunner().tableExists(getSession(), tableName));
+            assertUpdate("DROP TABLE " + tableName);
+        }
+        else {
+            String validTableName = baseTableName + "z".repeat(maxTableNameLength().getAsInt() - baseTableName.length());
+            assertUpdate("CREATE TABLE " + validTableName + " (a bigint)");
+            assertTrue(getQueryRunner().tableExists(getSession(), validTableName));
+            assertUpdate("DROP TABLE " + validTableName);
+
+            String invalidTableName = baseTableName + "z".repeat(maxTableNameLength().getAsInt() - baseTableName.length() + 1);
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2001]: invalidTableName = validTableName + "z";

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 4]
```diff
@@ -1978,6 +1978,38 @@ public void testCreateTable()
...
+            assertThatThrownBy(() -> query("CREATE TABLE " + invalidTableName + " (a bigint)"))
+                    .hasMessageMatching(".*"); // TODO: Check failed message
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2003]: remember to expect something reasonable, or let the connector handle that

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java [hunk 5]
```diff
@@ -472,6 +472,13 @@ protected String createSchemaSql(String schemaName)
         return "CREATE SCHEMA " + schemaName + " WITH (location = 's3://" + bucketName + "/" + schemaName + "')";
     }
 
+    @Override
+    public void testCreateTableWithLongTableName()
+    {
+        // TODO: Fails with java.net.SocketTimeoutException: Read timed out at ThriftHiveMetastore.createTable
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:478]: That sounds like a failure on metastore side (like HTTP 500) and it didn't return any response.

Do metastore logs say anything?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 6]
```diff
@@ -3700,6 +3701,13 @@ public void testModifyingOldSnapshotIsNotPossible()
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        // This value depends on file system
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:3707]: and not `// This value depends on HMS type` ?

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduConnectorTest.java [hunk 7]
```diff
@@ -276,6 +277,28 @@ public void testCreateTable()
         throw new SkipException("TODO");
     }
 
+    @Override
```
Comment (findepi) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduConnectorTest.java:280]: why overridden? please document

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 8]
```diff
@@ -206,7 +206,7 @@ public void createTable(SchemaTableName name, List<MongoColumnHandle> columns, O
             throw new SchemaNotFoundException(name.getSchemaName());
         }
         createTableMetadata(name, columns, comment);
-        // collection is created implicitly
```
Comment (findepi) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:209]: why this change?

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixClient.java [hunk 9]
```diff
@@ -652,6 +653,14 @@ public JdbcOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
         }
     }
 
+    @Override
+    protected void verifyTableNameLength(DatabaseMetaData databaseMetadata, String tableName)
```
Comment (findepi) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixClient.java:657]: What would happen if we did not validate anything here?
would Phoenix server reject the CREATE TABLE?

File: plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java [hunk 10]
```diff
@@ -651,6 +652,21 @@ protected void verifyConcurrentAddColumnFailurePermissible(Exception e)
...
+        // TODO: Find the maximum table name length in Phoenix and implement this test.
+        // Table name length with 65536 chars throws "startRow's length must be less than or equal to 32767 to meet the criteria for a row key."
+        // 32767 chars still causes the same error and shorter names (e.g. 10000) causes timeout.
+        throw new SkipException("TODO");
+    }
+
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        return OptionalInt.of(32767);
```
Comment (findepi) [inline plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:667]: remove this and `MAX_TABLE_NAME_LENGTH = 32767;` in the client  if the TODO above stays

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 8 above)
Comment (ebyhr) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:209]: The commit message has
> Let MongoDB creates a collection explicitly because 
implicit creation doesn't happen in case of long identifiers.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:560]: we need to do the check before sending the CREATE to PostgreSQL, otherwise we create a table with a wrong name.

for CREATE TABLE we can do the check in `io.trino.plugin.postgresql.PostgreSqlClient#createTable`.

is RENAME TO also affected?

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 8 above)
Comment (findepi) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:209]: Does `createTableMetadata(name, columns, comment` succeed?
does it create any resources?

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixClient.java (hunk 9 above)
Comment (ebyhr) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixClient.java:657]: Phoenix behavior is little tricky. I haven't investigated the border number. 
* `65536` chars: reject creation
* `32767` chars: cause timeout

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:560]: Yes, `RENAME TO` also truncates the name.

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixClient.java (hunk 9 above)
Comment (findepi) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixClient.java:657]: feel free to leave just a TODO for Phoenix and follow-up separately.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java (hunk 1 above)
Comment (findepi) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java:560]: that sounds like  we want to have a table validation hook in JdbcClient
and it would be called before CT and before Rename is issued to the remote system.

by default it would be empty, but for PostgreSQL we would add length checks.

**Comment by ebyhr** [comment]
CI failure seems related to #11344

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 11]
```diff
@@ -1978,6 +1978,43 @@ public void testCreateTable()
         assertFalse(getQueryRunner().tableExists(getSession(), tableNameLike));
     }
 
+    // TODO: Add RENAME TABLE test with long table name
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:1981]: Can be a follow up

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 12]
```diff
@@ -1978,6 +1978,43 @@ public void testCreateTable()
...
+                .orElse(65536 + 5);
+
+        String validTableName = baseTableName + "z".repeat(maxLength - baseTableName.length());
+        assertUpdate("CREATE TABLE " + validTableName + " (a bigint)");
+        assertTrue(getQueryRunner().tableExists(getSession(), validTableName));
+        assertUpdate("DROP TABLE " + validTableName);
+
+        String invalidTableName = validTableName + "z";
+        try {
+            assertUpdate("CREATE TABLE " + invalidTableName + " (a bigint)");
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2000]: we know invalid table name only when `maxTableNameLength()` is not empty.

(otherwise we wouldn't need to return Optional there, we'd just return `65536 + 5` in the default implementaion)

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorTest.java [hunk 13]
```diff
@@ -674,6 +674,25 @@ protected TestTable simpleTable()
...
+        assertTrue(getQueryRunner().tableExists(getSession(), validTableName));
+        assertThatThrownBy(() -> assertUpdate("DROP TABLE " + validTableName))
+                .hasMessageMatching("(?s).*(Bad path syntax|File name too long).*");
+
+        String invalidTableName = baseTableName + "z".repeat(maxTableNameLength().orElseThrow() - baseTableName.length() + 1);
+        assertThatThrownBy(() -> query("CREATE TABLE " + invalidTableName + " (a bigint)"))
+                .hasMessageMatching("(?s).*(Cannot open file|File name too long).*");
+        // ClickHouse lefts a table even if the above statement failed
+        assertTrue(getQueryRunner().tableExists(getSession(), validTableName));
+    }
```
Comment (findepi) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorTest.java:694]: add
```
assertThatThrownBy(() -> query("DROP TABLE ...
```

so that we know to update the test when clickhouse behavior is fixed

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java [hunk 14]
```diff
@@ -58,4 +60,11 @@ public void testCreateTableAsSelectWithTableComment()
         assertThatThrownBy(super::testCreateTableAsSelectWithTableComment)
                 .hasMessageMatching("(?s).* Syntax error: .* COMMENT 'test comment'.*");
     }
+
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        // The numeric value depends on file system
+        return OptionalInt.of(255 - ".sql.tmp".length());
```
Comment (findepi) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:68]: where comes `.sql.tmp` from?

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseLatestConnectorTest.java [hunk 15]
```diff
@@ -35,4 +37,11 @@ protected QueryRunner createQueryRunner()
                         .buildOrThrow(),
                 REQUIRED_TPCH_TABLES);
     }
+
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        // The numeric value depends on file system
+        return OptionalInt.of(255 - ".sql.detached".length());
```
Comment (findepi) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseLatestConnectorTest.java:45]: where comes `.sql.detached` from?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java [hunk 16]
```diff
@@ -472,6 +473,18 @@ protected String createSchemaSql(String schemaName)
...
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        return OptionalInt.of(128);
+    }
+
+    @Override
+    protected void verifyTableNameLengthFailurePermissible(Throwable e)
+    {
+        assertThat(e).hasMessageMatching("(?s)(.*Read timed out)|(.*\"`TBL_NAME`\" that has maximum length of 128.*)");
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:485]: Are you still getting `Read timed out` errors after specifying 128?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 17]
```diff
@@ -8661,6 +8662,19 @@ protected TestTable createTableWithDefaultColumns()
         throw new SkipException("Hive connector does not support column default values");
     }
 
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        // This value depends on HMS type
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8668]: HMS is Hive metastore, so it's a concrete type

```suggestion
        // This value depends on metastore type
```

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java [hunk 18]
```diff
@@ -8661,6 +8662,19 @@ protected TestTable createTableWithDefaultColumns()
...
+        return OptionalInt.of(255);
+    }
+
+    @Override
+    protected void verifyTableNameLengthFailurePermissible(Throwable e)
+    {
+        assertThat(e).hasMessageContaining("Failed to create directory");
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8675]: Is it same on Linux CI and Mac OS development box?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 19]
```diff
@@ -4274,6 +4275,19 @@ public void testCreateTableAsSelectFromVersionedTable()
         assertUpdate("DROP TABLE " + timestampVersionedSinkTableName);
     }
 
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        // This value depends on HMS type
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:4281]:
```suggestion
        // This value depends on metastore type
```

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduConnectorTest.java [hunk 20]
```diff
@@ -344,6 +344,29 @@ public void testCreateTable()
         //assertFalse(getQueryRunner().tableExists(getSession(), tableNameLike));
     }
 
+    @Override
+    public void testCreateTableWithLongTableName()
+    {
+        // DDL in base class can't create Kudu table due to lack of primary key and required table properties
```
Comment (findepi) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduConnectorTest.java:350]: "Overridden because ..."

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/BaseMongoConnectorTest.java [hunk 21]
```diff
@@ -576,6 +577,18 @@ public void testAddColumnConcurrently()
         throw new SkipException("TODO");
     }
 
+    @Override
+    protected OptionalInt maxTableNameLength()
+    {
+        return OptionalInt.of(120 - "test.".length());
```
Comment (findepi) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/BaseMongoConnectorTest.java:583]: where does "test." come from?
(the schema name is "tpch")

File: plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java (hunk 10 above)
Comment (findepi) [reply plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:679]: remove `maxTableNameLength()` implementation

File: plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java [hunk 22]
```diff
@@ -72,7 +73,7 @@ protected QueryRunner createQueryRunner()
             throws Exception
     {
         testingPhoenixServer = TestingPhoenixServer.getInstance();
-        return createPhoenixQueryRunner(testingPhoenixServer, ImmutableMap.of(), REQUIRED_TPCH_TABLES);
+        return createPhoenixQueryRunner(testingPhoenixServer, ImmutableMap.of(), ImmutableList.of());
```
Comment (findepi) [inline plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:76]: revert

**Review by findepi — APPROVED** [review]

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java (hunk 14 above)
Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseConnectorTest.java:68]: It comes from ClickHouse internal implementation.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseLatestConnectorTest.java (hunk 15 above)
Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestClickHouseLatestConnectorTest.java:45]: It comes from ClickHouse internal implementation.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorTest.java (hunk 13 above)
Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorTest.java:694]: The logic to capture future ClickHouse change already exists in L686.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java (hunk 18 above)
Comment (ebyhr) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/BaseHiveConnectorTest.java:8675]: Yes, it's same.
